internal/pkg/cli/env_upgrade.go

```diff
@@ -212,6 +224,44 @@ func (o *envUpgradeOpts) listEnvsToUpgrade() ([]*config.Environment, error) {
 	return envs, nil
 }
 
+func (o *envUpgradeOpts) ensureManagerRoleIsAllowedToUpload(env *config.Environment, bucketName string) error {
+	cfn, err := o.newEnvDeployer(env)
+	if err != nil {
+		return err
+	}
+	body, err := cfn.EnvironmentTemplate(env.App, env.Name)
+	if err != nil {
+		return err
+	}
+	if o.isManagerRoleAllowedToUpload(body, bucketName) {
+		return nil
+	}
+	// TODO(efekarakus): if the role isn't allowed to upload, modify the template to add the permission and call cfn.UpdateEnvironmentTemplate
+	return nil
+}
+
+func (o *envUpgradeOpts) isManagerRoleAllowedToUpload(template, bucketName string) bool {
+	wantedPolicyStatementID := "PutObjectsToArtifactBucket"
```

Comment:
Do you think we also want to check that the template version is < v1.18.0?

Reply:
I kept debating if we should just check if the Template version is < 1.9.0 and if so just add the permission, or go this route by checking specifically the values.

I can update it to use the Metadata.Version instead

Reply:
If we don't check the version, false positive could happen if in the we rearrange the policy statements a little, such that `Put` permissions are grouped into another statement and that we delete `PutArtifactsToS3`, or if we simply renamed the statement. I imagine this is not a big deal though.

(On a side note, I think false negative won't happen unless users have manual modifications to the template that happen to break the assumptions that support this implementation so we're good.)